packages/common/src/common.ts

```diff
@@ -285,57 +285,89 @@ export class Common extends EventEmitter {
+    let hardfork: HardforkConfig | undefined = undefined
+
+    // Hardforks specified are strickly in the order of application, so we can make some intelligent decisions
+    // based on that especially regarding the merge hardfork where block might not be specified
     for (const hf of this.hardforks()) {
-      // Skip comparison for not applied HFs
-      if (hf.block === null) {
-        if (td !== undefined && td !== null && hf.ttd !== undefined && hf.ttd !== null) {
-          if (td >= BigInt(hf.ttd)) {
-            return hf.name
-          }
-        }
+      // Check if hf can be applied, if not then break and the last hardfork set is the response
+      if (hf.block === null && (hf.ttd === undefined || hf.ttd === null)) {
+        // Not a hardfork which has been scheduled yet so just skip
         continue
       }
-      if (blockNumber >= BigInt(hf.block)) {
-        hardfork = hf.name as Hardfork
-      }
-      if (td && (typeof hf.ttd === 'string' || typeof hf.ttd === 'bigint')) {
-        if (td >= BigInt(hf.ttd)) {
-          minTdHF = hf.name
-        } else {
-          maxTdHF = previousHF
+
+      if (hf.ttd !== undefined && hf.ttd !== null) {
+        // Ok, this is the hardfork with ttd
+        // If preMergeHF assigned, we have already seen merge hardfork
+        if (preMergeHF !== undefined) {
+          throw new Error(
+            `Invalid hardfork config with repeat ttd hardfork=${hf.name} ttd=${hf.ttd}`
+          )
         }
+        preMergeHF = hardfork
       }
-      previousHF = hf.name
-    }
-    if (td) {
-      let msgAdd = `block number: ${blockNumber} (-> ${hardfork}), `
-      if (minTdHF !== undefined) {
-        if (!this.hardforkGteHardfork(hardfork, minTdHF)) {
-          const msg = 'HF determined by block number is lower than the minimum total difficulty HF'
-          msgAdd += `total difficulty: ${td} (-> ${minTdHF})`
-          throw new Error(`${msg}: ${msgAdd}`)
+
+      // If merge hardfork has block null set, we will make determination if this is merge hardfork or preMergeHF
+      //  - if no future hardforks qualify by block number
+      // So we assume it to qualify for now to move to check future post merge hardfork qualitfication
+      //
+      // For the hardforks with block specified we check for qualification
+      if (hf.block !== null) {
+        if (blockNumber < BigInt(hf.block)) {
+          break
         }
       }
-      if (maxTdHF !== undefined) {
-        if (!this.hardforkGteHardfork(maxTdHF, hardfork)) {
-          const msg = 'Maximum HF determined by total difficulty is lower than the block number HF'
-          msgAdd += `total difficulty: ${td} (-> ${maxTdHF})`
-          throw new Error(`${msg}: ${msgAdd}`)
+
+      // hf qualifies else we would have broken out of the loop till now
+      if (hardfork?.ttd !== undefined && hardfork?.ttd !== null) {
+        passedMergeHF = hardfork!
```

Comment:
Why do we need this in addition to `preMergeHf`?  I think I must not be tracking all of the logic yet because there can be only one hardfork with TTD set.  So, if we have that, can't we just use that hardfork (set above as `preMergeHf`).  Also, it seems really odd to call the merge hardfork `preMergeHf`.  The "pre-merge hardfork" is London so maybe we can find a better way to name this?

Reply:
yes, thats what `preMergeHf` should represent, its `london`, and `passedMergedHF` is assigned `merge` hardfork if its passed, i guess i can change it to boolean and store the merge hardfork as `mergeHF` as this later needs to be accessed.